internal/database/repository/application.go

```diff
@@ -51,62 +48,240 @@ func applyApplicationFilters(builder *sql.Builder, filters ApplicationFilters) {
+	`
+
+	_, err := s.dbpool.Exec(
+		ctx,
+		q,
+		pgx.NamedArgs{
+			"id":                   app.ID,
+			"created_at":           app.CreatedAt,
+			"deleted_at":           app.DeletedAt,
+			"app_id":               app.ApplicationID,
+			"used_resource":        app.UsedResource,
+			"max_used_resource":    app.MaxUsedResource,
+			"pending_resource":     app.PendingResource,
+			"partition":            app.Partition,
+			"queue_name":           app.QueueName,
+			"submission_time":      app.SubmissionTime,
+			"finished_time":        app.FinishedTime,
+			"requests":             app.Requests,
+			"allocations":          app.Allocations,
+			"state":                app.State,
+			"user":                 app.User,
+			"groups":               app.Groups,
+			"rejected_message":     app.RejectedMessage,
+			"state_log":            app.StateLog,
+			"place_holder_data":    app.PlaceholderData,
+			"has_reserved":         app.HasReserved,
+			"reservations":         app.Reservations,
+			"max_request_priority": app.MaxRequestPriority,
+		})
+	return err
+}
+
+func (s *PostgresRepository) GetLatestApplicationByApplicationID(ctx context.Context, appID string) (*model.Application, error) {
+	const q = `
+SELECT
+	id,
+	created_at,
+	deleted_at,
+	app_id,
+	used_resource,
+	max_used_resource,
+	pending_resource,
+	partition,
+	queue_name,
+	submission_time,
+	finished_time,
+	requests,
+	allocations,
+	state,
+	"user",
+	groups,
+	rejected_message,
+	state_log,
+	place_holder_data,
+	has_reserved,
+	reservations,
+	max_request_priority
+FROM
+	applications
+WHERE
+	app_id = @app_id
+ORDER BY id DESC
+LIMIT 1
+	`
+
+	var app model.Application
+	row := s.dbpool.QueryRow(
+		ctx,
+		q,
+		pgx.NamedArgs{
+			"app_id": appID,
+		},
+	)
+
+	if err := row.Scan(
+		&app.ID,
+		&app.CreatedAt,
+		&app.DeletedAt,
+		&app.ApplicationID,
+		&app.UsedResource,
+		&app.MaxUsedResource,
+		&app.PendingResource,
+		&app.Partition,
+		&app.QueueName,
+		&app.SubmissionTime,
+		&app.FinishedTime,
+		&app.Requests,
+		&app.Allocations,
+		&app.State,
+		&app.User,
+		&app.Groups,
+		&app.RejectedMessage,
+		&app.StateLog,
+		&app.PlaceholderData,
+		&app.HasReserved,
+		&app.Reservations,
+		&app.MaxRequestPriority,
+	); err != nil {
+		return nil, err
+	}
+
+	return &app, nil
+}
+
+func (s *PostgresRepository) GetLatestApplicationsByApplicationID(ctx context.Context) ([]*model.Application, error) {
+	const q = `
+SELECT DISTINCT ON (app_id)
+	id,
+	created_at,
+	deleted_at,
+	app_id,
+	used_resource,
+	max_used_resource,
+	pending_resource,
+	partition,
+	queue_name,
+	submission_time,
+	finished_time,
+	requests,
+	allocations,
+	state,
+	"user",
+	groups,
+	rejected_message,
+	state_log,
+	place_holder_data,
+	has_reserved,
+	reservations,
+	max_request_priority
+FROM
+	applications
+ORDER BY app_id, id DESC
```

Comment:
Just a quick question: Why do we need to order by `app_id`? Shouldn't the `id` alone be sufficient to keep the active ones at the top?

Reply:
Good question and no, it wouldn't be sufficient.
Imagine this scenario:
I create app with `id=1`. The app finishes and it is removed. Then, someone creates the app with `id=1`. Since there is no app in the scheduler with `id=1`, it is fine from the yunikorn perspective. But this is completely new app. We would get create event to insert the new app.
In yunikorn, app_id is unique, since the app is going to be removed after it completes. But for yhs, app ID doesn't have to be unique. If it were the case, then we wouldn't need our own ID field.
What I'm trying to do is get the latest version of the application with this app_id.